pkg/istio-agent/agent_test.go

```diff
@@ -331,6 +333,37 @@ func TestAgent(t *testing.T) {
 
 		a.Check(security.WorkloadKeyCertResourceName, security.RootCertReqResourceName)
 	})
+	t.Run("Envoy lifecycle", func(t *testing.T) {
```

Comment:
Running envoy here (or just more of the agent) seems to introduce a few races and failures
```
==================
WARNING: DATA RACE
Read at 0x00c000aeb248 by goroutine 31:
  istio.io/istio/pkg/istio-agent.(*Agent).Close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:428 +0x4a
  istio.io/istio/pkg/istio-agent.(*Agent).Close-fm()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:427 +0x4a
  testing.(*common).Cleanup.func1()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:902 +0x1be
  testing.(*common).runCleanup()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1008 +0xae
  testing.tRunner.func2()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1188 +0x74
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1198 +0x22a

Previous write at 0x00c000aeb248 by goroutine 27:
  istio.io/istio/pkg/istio-agent.(*Agent).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:366 +0xc5c
  istio.io/istio/pkg/istio-agent.Setup.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:445 +0x50

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1239 +0x5d7
  istio.io/istio/pkg/istio-agent.TestAgent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:119 +0x445
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202

Goroutine 27 (finished) created at:
  istio.io/istio/pkg/istio-agent.Setup()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:443 +0x1768
  istio.io/istio/pkg/istio-agent.TestAgent.func5()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:121 +0xc4
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000a99040 by goroutine 31:
  istio.io/istio/pkg/istio-agent.(*XdsProxy).close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:555 +0x3e
  istio.io/istio/pkg/istio-agent.(*Agent).Close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:429 +0x187
  istio.io/istio/pkg/istio-agent.(*Agent).Close-fm()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:427 +0x4a
  testing.(*common).Cleanup.func1()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:902 +0x1be
  testing.(*common).runCleanup()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1008 +0xae
  testing.tRunner.func2()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1188 +0x74
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1198 +0x22a

Previous write at 0x00c000a99040 by goroutine 27:
  istio.io/istio/pkg/istio-agent.initXdsProxy()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:140 +0x384
  istio.io/istio/pkg/istio-agent.(*Agent).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:366 +0xc15
  istio.io/istio/pkg/istio-agent.Setup.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:445 +0x50

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1239 +0x5d7
  istio.io/istio/pkg/istio-agent.TestAgent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:119 +0x445
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202

Goroutine 27 (finished) created at:
  istio.io/istio/pkg/istio-agent.Setup()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:443 +0x1768
  istio.io/istio/pkg/istio-agent.TestAgent.func5()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:121 +0xc4
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000a99130 by goroutine 31:
  istio.io/istio/pkg/istio-agent.(*XdsProxy).close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:556 +0x64
  istio.io/istio/pkg/istio-agent.(*Agent).Close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:429 +0x187
  istio.io/istio/pkg/istio-agent.(*Agent).Close-fm()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:427 +0x4a
  testing.(*common).Cleanup.func1()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:902 +0x1be
  testing.(*common).runCleanup()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1008 +0xae
  testing.tRunner.func2()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1188 +0x74
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1198 +0x22a

Previous write at 0x00c000a99130 by goroutine 27:
  istio.io/istio/pkg/istio-agent.initXdsProxy()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:140 +0x384
  istio.io/istio/pkg/istio-agent.(*Agent).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:366 +0xc15
  istio.io/istio/pkg/istio-agent.Setup.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:445 +0x50

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1239 +0x5d7
  istio.io/istio/pkg/istio-agent.TestAgent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:119 +0x445
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202

Goroutine 27 (finished) created at:
  istio.io/istio/pkg/istio-agent.Setup()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:443 +0x1768
  istio.io/istio/pkg/istio-agent.TestAgent.func5()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:121 +0xc4
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c0010aebb8 by goroutine 31:
  istio.io/istio/pkg/wasm.(*LocalFileCache).Cleanup()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/wasm/cache.go:153 +0x3e
  istio.io/istio/pkg/istio-agent.(*XdsProxy).close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:556 +0x81
  istio.io/istio/pkg/istio-agent.(*Agent).Close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:429 +0x187
  istio.io/istio/pkg/istio-agent.(*Agent).Close-fm()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:427 +0x4a
  testing.(*common).Cleanup.func1()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:902 +0x1be
  testing.(*common).runCleanup()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1008 +0xae
  testing.tRunner.func2()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1188 +0x74
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1198 +0x22a

Previous write at 0x00c0010aebb8 by goroutine 27:
  istio.io/istio/pkg/wasm.NewLocalFileCache()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/wasm/cache.go:87 +0x144
  istio.io/istio/pkg/istio-agent.initXdsProxy()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:148 +0x349
  istio.io/istio/pkg/istio-agent.(*Agent).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:366 +0xc15
  istio.io/istio/pkg/istio-agent.Setup.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:445 +0x50

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1239 +0x5d7
  istio.io/istio/pkg/istio-agent.TestAgent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:119 +0x445
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202

Goroutine 27 (finished) created at:
  istio.io/istio/pkg/istio-agent.Setup()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:443 +0x1768
  istio.io/istio/pkg/istio-agent.TestAgent.func5()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:121 +0xc4
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000a990d8 by goroutine 31:
  istio.io/istio/pkg/istio-agent.(*XdsProxy).close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:557 +0x9b
  istio.io/istio/pkg/istio-agent.(*Agent).Close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:429 +0x187
  istio.io/istio/pkg/istio-agent.(*Agent).Close-fm()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:427 +0x4a
  testing.(*common).Cleanup.func1()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:902 +0x1be
  testing.(*common).runCleanup()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1008 +0xae
  testing.tRunner.func2()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1188 +0x74
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1198 +0x22a

Previous write at 0x00c000a990d8 by goroutine 27:
  istio.io/istio/pkg/istio-agent.initXdsProxy()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:140 +0x384
  istio.io/istio/pkg/istio-agent.(*Agent).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:366 +0xc15
  istio.io/istio/pkg/istio-agent.Setup.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:445 +0x50

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1239 +0x5d7
  istio.io/istio/pkg/istio-agent.TestAgent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:119 +0x445
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202

Goroutine 27 (finished) created at:
  istio.io/istio/pkg/istio-agent.Setup()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:443 +0x1768
  istio.io/istio/pkg/istio-agent.TestAgent.func5()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:121 +0xc4
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Write at 0x00c0005af038 by goroutine 31:
  sync/atomic.StoreInt32()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/runtime/race_amd64.s:242 +0xb
  net/http.(*atomicBool).setTrue()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/net/http/server.go:527 +0x6e
  net/http.(*Server).Close()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/net/http/server.go:2689 +0x50
  istio.io/istio/pkg/istio-agent.(*XdsProxy).close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:558 +0x171
  istio.io/istio/pkg/istio-agent.(*Agent).Close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:429 +0x187
  istio.io/istio/pkg/istio-agent.(*Agent).Close-fm()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:427 +0x4a
  testing.(*common).Cleanup.func1()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:902 +0x1be
  testing.(*common).runCleanup()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1008 +0xae
  testing.tRunner.func2()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1188 +0x74
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1198 +0x22a

Previous write at 0x00c0005af038 by goroutine 27:
  istio.io/istio/pkg/istio-agent.(*XdsProxy).initDebugInterface()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:810 +0x18a
  istio.io/istio/pkg/istio-agent.(*Agent).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:371 +0xe0e
  istio.io/istio/pkg/istio-agent.Setup.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:445 +0x50

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1239 +0x5d7
  istio.io/istio/pkg/istio-agent.TestAgent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:119 +0x445
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202

Goroutine 27 (finished) created at:
  istio.io/istio/pkg/istio-agent.Setup()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:443 +0x1768
  istio.io/istio/pkg/istio-agent.TestAgent.func5()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:121 +0xc4
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Write at 0x00c0005af060 by goroutine 31:
  sync/atomic.CompareAndSwapInt32()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/runtime/race_amd64.s:321 +0xb
  sync.(*Mutex).Lock()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/sync/mutex.go:74 +0x4d
  net/http.(*Server).Close()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/net/http/server.go:2690 +0x8e
  istio.io/istio/pkg/istio-agent.(*XdsProxy).close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:558 +0x171
  istio.io/istio/pkg/istio-agent.(*Agent).Close()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:429 +0x187
  istio.io/istio/pkg/istio-agent.(*Agent).Close-fm()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:427 +0x4a
  testing.(*common).Cleanup.func1()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:902 +0x1be
  testing.(*common).runCleanup()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1008 +0xae
  testing.tRunner.func2()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1188 +0x74
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1198 +0x22a

Previous write at 0x00c0005af060 by goroutine 27:
  istio.io/istio/pkg/istio-agent.(*XdsProxy).initDebugInterface()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/xds_proxy.go:810 +0x18a
  istio.io/istio/pkg/istio-agent.(*Agent).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent.go:371 +0xe0e
  istio.io/istio/pkg/istio-agent.Setup.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:445 +0x50

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1239 +0x5d7
  istio.io/istio/pkg/istio-agent.TestAgent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:119 +0x445
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202

Goroutine 27 (finished) created at:
  istio.io/istio/pkg/istio-agent.Setup()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:443 +0x1768
  istio.io/istio/pkg/istio-agent.TestAgent.func5()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/istio-agent/agent_test.go:121 +0xc4
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.16/src/testing/testing.go:1194 +0x202
==================
2021-05-13T15:40:36.146812Z     error   error serving tap http server: http: Server closed
2021-05-13T15:40:36.147715Z     warn    ca      ca request failed, starting attempt 1 in 97.465172ms
2021-05-13T15:40:36.147924Z     info    kube    Controller terminated
    testing.go:1093: race detected during execution of test
```

Reply:
I suspect it's because we run agent in one routine (because it's blocking) and then close in another. I think we should defer closure in the routine that runs it.

Reply:
Refactored to use wait groups. I think other side functions also need to join the group, but that's not needed in this PR.